File: utils/vex2ccf.cc
#include <Vex++.h>

#include <assert.h>
#include <string>
#include <stdio.h>
#include <fstream>
#include <set>

#include <json/json.h>

std::string add_time(std::string &time, int delta) {
  int year, day, hour, minute, second;
  int n_matched = sscanf(time.c_str(), "%dy%dd%dh%dm%ds", 
                         &year, &day, &hour, &minute, &second);
  assert(n_matched == 5);
  second += delta;
  minute += second/60;
  hour += minute/60;
  day  += hour/24;
  int days_per_year = 365;
  if ((year % 4 == 0) && ((year % 100 != 0) && (year % 400 == 0))) {
    days_per_year = 366;
  }
  year += day/days_per_year;
  char result[20];
  snprintf(result, 20, "%04dy%03dd%02dh%02dm%02ds",
           year, (day%days_per_year), (hour%24), (minute%60), (second%60));
  return std::string(result);
}

std::string get_start(const Vexpp_node &vex) {
  std::string result = "";
  Vex::Node::const_iterator sched = vex["SCHED"];
  for (Vex::Node::const_iterator scan_it = sched->begin();
       scan_it != sched->end(); ++scan_it) {
    if (result == "") {
      result = scan_it["start"]->to_string();
    } else if (result > scan_it["start"]->to_string()) {
      result = scan_it["start"]->to_string();
    }
  }
  return result;
}

std::string get_endtime_of_scan(const Vexpp_node &scan) {
  std::string start = scan["start"]->to_string();
  int duration = 0;
  for (Vex::Node::const_iterator it = scan.begin("station");
       it != scan.end("station"); ++it) {
    int curr_duration;
    sscanf(it[2]->to_string().c_str(),"%d sec", &curr_duration);
    duration = (curr_duration > duration ? curr_duration : duration);
  }
  return add_time(start, duration);
}

std::string get_stop(const Vexpp_node &vex) {
  std::string result = "";
  Vex::Node::const_iterator sched = vex["SCHED"];
  for (Vex::Node::const_iterator scan_it = sched->begin();
       scan_it != sched->end(); ++scan_it) {
    std::string curr = get_endtime_of_scan(*scan_it);
    if (result == "") {
      result = curr;
    } else if (result < curr) {
      result = curr;
    }
  }
  return result;
}

Json::Value get_frequency(const Vex &vex,
                          Vexpp_node::const_iterator channel,
                          const std::string &ref_station, 
                          const std::string &BBC_block,
                          const std::string &IF_block) {
  Json::Value result;
  std::string if_ref;
  for (Vex::Node::const_iterator bbc = 
         vex.get_root_node()["BBC"][BBC_block]->begin("BBC_assign");
       bbc != vex.get_root_node()["BBC"][BBC_block]->end("BBC_assign"); ++bbc) {
    if (bbc[0]->to_string() == channel[5]->to_string()) {
      if_ref = bbc[2]->to_string();
    }
  }

  double freq, bandwidth;
  int err = sscanf(channel[1]->to_string().c_str(), "%lf MHz", &freq);
  assert(err == 1);
  err = sscanf(channel[3]->to_string().c_str(), "%lf MHz", &bandwidth);
  assert(err == 1);
  
  result["frequency"] = freq*1000000;
  result["bandwidth"] = bandwidth*1000000;
  result["sideband"]  = channel[2]->to_string()+"SB";
  result["polarisation"]  = std::string(1,vex.polarisation(IF_block, if_ref));
  return result;
}

Json::Value get_frequencies(const Vex &vex) {
  // Find the right BBC and IF
  std::string mode = vex.get_root_node()["MODE"]->begin().key();
  std::string IF_block = vex.get_root_node()["MODE"][mode]["IF"][0]->to_string();
  std::string ref_station = vex.get_root_node()["MODE"][mode]["IF"][1]->to_string();
  std::string BBC_block = "";
  // Find the corresponding BBC block:
  for (Vex::Node::const_iterator bbc = vex.get_root_node()["MODE"][mode]->begin("BBC");
       bbc != vex.get_root_node()["MODE"][mode]->end("BBC"); ++bbc) {
    for (Vex::Node::const_iterator station = ++(bbc->begin()); 
	 station != (*bbc).end(); ++station) {
      if (ref_station == station->to_string()) {
	BBC_block = bbc[0]->to_string();
      }
    }
  }

  Json::Value result;
  for (Vex::Node::const_iterator channel = vex.get_root_node()["FREQ"]->begin()->begin("chan_def");
       channel != vex.get_root_node()["FREQ"]->begin()->end("chan_def"); ++channel) {
    result.append(get_frequency(vex, channel, ref_station, BBC_block, IF_block));
  }
//   result = []
//   for freq in vex["FREQ"]:
//     for channel in vex["FREQ"][freq].getall("chan_def"):
//       result.append(get_frequency(IF, BBC, channel))
//     break
  return result;
}

Json::Value site_position(const Vexpp_node &vex,
			  const std::string &station) {
  Json::Value result;

  Vex::Node::const_iterator position = 
    vex["SITE"][vex["STATION"][station]["SITE"]->to_string()]["site_position"];
  for (Vex::Node::const_iterator site = position->begin();
       site != position->end(); ++site) {
    double pos;
    int err = sscanf(site->to_string().c_str(), "%lf m", &pos);
    assert(err == 1);
    result.append(pos);
  }
  return result;
}

Json::Value get_channels(const Vex &vex){
	std::set<std::string> result_set;
	Json::Value result;
	
  for (Vex::Node::const_iterator frq_block = vex.get_root_node()["FREQ"]->begin();
       frq_block != vex.get_root_node()["FREQ"]->end(); ++frq_block) {
  	for (Vex::Node::const_iterator freq_it = frq_block->begin("chan_def");
         freq_it != frq_block->end("chan_def"); ++freq_it) {
			result_set.insert(freq_it[4]->to_string());
		}
	}
	for (std::set<std::string>::const_iterator set_it = result_set.begin();
				set_it != result_set.end(); ++set_it){
		result.append(*set_it);
	}
	return result;
}


int main(int argc, char *argv[]) {
  assert(argc == 3);

	std::ofstream outfile(argv[2], std::ios::out);

  Vex vex(argv[1]);
  
  Json::Value json_output;
  json_output["exper_name"] = 
    vex.get_root_node()["GLOBAL"]["EXPER"]->to_string();
  json_output["start"] = get_start(vex.get_root_node());
  json_output["stop"] = get_stop(vex.get_root_node());
  for (Vex::Node::const_iterator it = vex.get_root_node()["STATION"]->begin();
       it != vex.get_root_node()["STATION"]->end(); ++it) {
    json_output["stations"].append(it.key());
    json_output["data_sources"][it.key()] = Json::Value(Json::arrayValue);
    json_output["site_position"][it.key()] = site_position(vex.get_root_node(), it.key());
  }

	json_output["channels"] =  get_channels(vex);
  json_output["reference_station"] = "";
  json_output["cross_polarize"]    = false;
  json_output["number_channels"]   = 1024;
  json_output["integr_time"]       = 1;
  json_output["message_level"]     = 0;
  json_output["delay_directory"]   = "";
  json_output["output_file"]       = "";
  json_output["subbands"]          = get_frequencies(vex);

  outfile << json_output << std::endl;
  return 0;
} 

File: webservices/vex2ccf_web_service/vex2ccf.py
#!/usr/bin/python

# 2007, Developed by Nico Kruithof <Kruithof@jive.nl>

# 2009-05-27 (iii) small:
# Multidict has a tendency to throw away duplicate keys
# so it is especially useless for locating channels;
# I've papered over with grep, but really we should use a 
# _real_ vex parser

import sys, re, os
#from vex_parser import Vex
from Vex import Vex
import simplejson, tempfile
import pprint

def add_time(time, delta):
    start_date = re.match("(\d*)y(\d*)d(\d*)h(\d*)m(\d*)s", time).groups()
    second = int(start_date[4]) + int(delta)
    minute = int(start_date[3]) + (second / 60)
    hour   = int(start_date[2]) + (minute / 60)
    day    = int(start_date[1]) + (hour   / 24)
    year   = int(start_date[0])
    days_per_year = 365
    if (year % 4 == 0) and ((year % 100 != 0) or (year % 400 == 0)):
        days_per_year = 366
        year = year + (day / days_per_year)
    return "%04dy%03dd%02dh%02dm%02ds" % \
        (year, (day%days_per_year), (hour%24), (minute%60), (second%60))

def get_start(vex):
    start = ""
    sched = vex["SCHED"]
    for scan in sched:
        if start == "":
            start = sched[scan]["start"]
    else:
        if start > sched[scan]["start"]:
            start = sched[scan]["start"]
    return start

def get_scan_end(scan):
    return add_time(scan["start"], 
                        max([i[2].split()[0] for i in scan.getall("station")]))

def get_stop(vex):
    sched = vex["SCHED"]
    endtimes = [get_scan_end(sched[scan]) for scan in sched]
    return max(endtimes)

def get_scan_ranges(vex):
    sched = vex["SCHED"]
    return [(sched[scan]['start'], get_scan_end(sched[scan])) for scan in sched]

def get_frequency(vex, IF, BBC, channel):
    if_def = ""
    for bbc in vex["BBC"][BBC].getall("BBC_assign"):
        if bbc[0] == channel[5]:
            if_def = bbc[2]
            
            result = dict()
            result["frequency"]    = float(channel[1].split()[0])*1000000
            result["bandwidth"]    = float(channel[3].split()[0])*1000000
            result["sideband"]     = channel[2]+"SB"
            for if_def_it in vex["IF"][IF].getall("if_def"):
                if if_def_it[0] == if_def:
                    result["polarisation"] = if_def_it[2]
    return result

def get_frequencies(vex):
    # take the first frequency, even if there are multiple frequencies
    for i in vex["MODE"]: mode = i
    IF = vex["MODE"][mode]["IF"][0]
    ref_station = vex["MODE"][mode]["IF"][1]
    BBC = ""
    ## Find the corresponding BBC:
    for bbc in vex["MODE"][mode].getall("BBC"):
        for station in bbc[1:]:
            if station == ref_station:
                BBC = bbc[0]

                result = []
                for freq in vex["FREQ"]:
                    for channel in vex["FREQ"][freq].getall("chan_def"):
                        result.append(get_frequency(vex, IF, BBC, channel))
                break
    return result


def vex2dict(vex):
    json_output = dict()
    json_output["exper_name"] = vex["GLOBAL"]["EXPER"]
    json_output["start"] = get_start(vex)
    json_output["stop"] = get_stop(vex)
    json_output["scans"] = get_scan_ranges(vex)
    json_output["reference_station"] = ""
    json_output["cross_polarize"]    = False
    json_output["number_channels"]   = 1024
    json_output["subbands"] = get_frequencies(vex)
    json_output["integr_time"]       = 1.0
    json_output["message_level"]     = 1
    json_output["delay_directory"]   = ""
    json_output["output_file"]       = ""
    json_output["data_sources"]      = dict()
    json_output["site_position"]     = dict()
    json_output["stations"]          = []
    for station in vex["STATION"]:
        json_output["stations"].append(station)
        json_output["data_sources"][station] = []
        site = vex["STATION"][station]["SITE"]
        json_output["site_position"][station] = [float(pos.split()[0])
                                                 for pos in 
                                                 vex["SITE"][site]["site_position"] ]
    return json_output


import popen2
def vex2ccf(vex_fn):
    vex = Vex(vex_fn)
    pout, pin = popen2.popen2("grep -o '&CH[0-9]*' %s | sort | uniq" % vex_fn)
    channels = [l.strip() for l in pout.readlines()]
    vexdict = vex2dict(vex)
    vexdict['channels'] = channels
    return simplejson.dumps(vexdict)

from vex2ccf_services import *

def convertRequest(vex_content):
    print >>sys.stderr, "Got request"
    vex = vex_content['vex_file']
    tfn = tempfile.mktemp()
    tf = file(tfn, 'w')
    tf.write(vex)
    tf.write('\n')
    tf.close()
    print >>sys.stderr, "************************************"
    s = vex2ccf(tfn)
    os.unlink(tfn)
    response = convertResponse()
    response._ccf_file = unicode(s)
    return response

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print "%s takes a vex-file as argument" % sys.argv[0]
        sys.exit(1)
    f = file(sys.argv[1])
    #s = f.read()
    #print vex2ccf(s)
    print vex2ccf(sys.argv[1])
